### PR TITLE
Enhancement: Allowing zero length in static array

### PR DESCRIPTION
**Pull request by ahangsu**
Fixes https://github.com/algorand/js-algorand-sdk/issues/497

### CODE REVIEW

File: src/main/java/com/algorand/algosdk/abi/TypeArrayStatic.java [hunk 1]
```diff
@@ -7,8 +7,8 @@ public class TypeArrayStatic extends ABIType {
     public final int length;
 
     public TypeArrayStatic(ABIType elemType, int length) {
-        if (length < 1)
-            throw new IllegalArgumentException("static-array initialize failure: array length should be at least 1");
+        if (length < 0)
+            throw new IllegalArgumentException("static-array initialize failure: array length should be positive");
```
Comment (bbroder-algo) [inline src/main/java/com/algorand/algosdk/abi/TypeArrayStatic.java:11]: non-negative

File: .test-env [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # Configs for testing repo download:
 SDK_TESTING_URL="https://github.com/algorand/algorand-sdk-testing"
-SDK_TESTING_BRANCH="master"
+SDK_TESTING_BRANCH="zero-length"
```
[the remark this answers is not in the document]
Comment (ahangsu) [reply .test-env:3]: yes, we should remember reverting back to master before merge!
```suggestion
SDK_TESTING_BRANCH="master"
```

File: src/main/java/com/algorand/algosdk/abi/TypeArrayStatic.java (hunk 1 above)
Comment (ahangsu) [reply src/main/java/com/algorand/algosdk/abi/TypeArrayStatic.java]: 👍 good find

File: .test-env (hunk 2 above)
[the remark this answers is not in the document]
Comment (ahangsu) [reply .test-env:3]: The new testcase (cucumber test) is added here: https://github.com/algorand/algorand-sdk-testing/pull/250/files

**Review by michaeldiamant — APPROVED** [review]
@ahangsu Thanks for bringing the fix here!

File: .test-env (hunk 2 above)
[the remark this answers is not in the document]
Comment (michaeldiamant) [reply .test-env:3]: > is this supposed to be checked in?

@bbroder-algo To add a bit more context and happy to discuss live - We opt to commit the `.test-env` config to ensure everyone can reproduce the branch's behavior.  We've previously run into difficulties debugging due to varied local `.test-env` configs.

File: src/main/java/com/algorand/algosdk/abi/ABIType.java [hunk 3]
```diff
@@ -9,7 +9,7 @@
 
 public abstract class ABIType {
     public static final int ABI_DYNAMIC_HEAD_BYTE_LEN = 2;
-    private static final Pattern staticArrayPattern = Pattern.compile("^(?<elemT>[a-z\\d\\[\\](),]+)\\[(?<len>[1-9][\\d]*)]$");
+    private static final Pattern staticArrayPattern = Pattern.compile("^(?<elemT>[a-z\\d\\[\\](),]+)\\[(?<len>0|[1-9][\\d]*)]$");
```
[the remark this answers is not in the document]
Comment (bbroder-algo) [reply src/main/java/com/algorand/algosdk/abi/ABIType.java:12]: nevermind. reread the arc.

File: src/test/java/com/algorand/algosdk/abi/TestTypes.java [hunk 4]
```diff
@@ -314,6 +314,7 @@ public void TestTypeFromStringInvalid() {
                 "[][][]",
                 "stuff[]",
                 // static array
+                "bool[01]",
```
Comment (bbroder-algo) [inline src/test/java/com/algorand/algosdk/abi/TestTypes.java:317]: nice

**Review by algochoi — APPROVED** [review]

**Pull request merged**